src/VirtualDeskManager.cpp

```diff
@@ -252,24 +252,66 @@ void VirtualDeskManager::resetVdesk(const std::string& arg) {
     vdesksMap[vdeskId]->resetLayout();
 }
 
+// Returns true if any workspace in the vdesk has at least one window
+bool VirtualDeskManager::isDeskPopulated(int vdeskId) {
+    if (!vdesksMap.contains(vdeskId))
+        return false;
+    auto vdesk = vdesksMap[vdeskId];
+    for (const auto& [monitor, workspaceId] : vdesk->activeLayout(conf)) {
+        auto workspace = g_pCompositor->getWorkspaceByID(workspaceId);
+        if (workspace && workspace->getWindows() > 0)
+            return true;
+    }
+    return false;
+}
+
 int VirtualDeskManager::prevDeskId(bool backwardCycle) {
-    int prevId = activeVdesk()->id - 1;
-    if (prevId < 1) {
-        prevId = 1;
-        if (backwardCycle) {
-            auto keys = std::views::keys(vdesksMap);
-            prevId    = std::ranges::max(keys);
+    // Read config: cycle_populated_only
+    static auto* const PCYCLE_POPULATED_ONLY = (Hyprlang::INT* const*)HyprlandAPI::getConfigValue(PHANDLE, CYCLE_POPULATED_ONLY_CONF)->getDataStaticPtr();
+    bool populatedOnly = **PCYCLE_POPULATED_ONLY;
+
+    int currentId = activeVdesk()->id;
+    int candidateId = currentId;
+    auto keys = std::views::keys(vdesksMap);
+    int minId = std::ranges::min(keys);
+    int maxId = std::ranges::max(keys);
+    do {
+        candidateId--;
+        if (candidateId < minId) {
+            if (backwardCycle)
+                candidateId = maxId;
+            else
+                return minId;
         }
-    }
-    return prevId;
+        // A desk must exist to be a candidate
+        if (vdesksMap.contains(candidateId) && (!populatedOnly || isDeskPopulated(candidateId)))
+            return candidateId;
+    } while (candidateId != currentId);
+    return currentId;
 }
 
 int VirtualDeskManager::nextDeskId(bool cycle) {
```

Comment:
same comment as for the other function, here you need to keep the vdesksmap.contains call though. In general, though, there's too much repeated code between the two functions, maybe we could simplify this?